### PR TITLE
1918 Upper slim validation and coverage report

### PR DESCRIPTION
**Pull request by ubyndr**
Addresses #1918, Extend CL ODK with infrastructure for validating upper slims and generating coverage reports.

- Added a python script for creating upper slim validation and coverage reports.
- Extended cl.Makefile with `UPPER SLIM VALIDATION AND COVERAGE REPORTS` section

> * QC: Fail test if any term in upper_slim is a subclass of another term in the same upper_slim; (enforce naming convention? regex .+_upper_slim$ ???)

I've added `slim_coverage` target to `test` target for QC github action. Following exception is thrown when an upper slim fails the validation;
```
Exception: ../templates/blood_and_immune_upper_slim.csv is invalid! macrophage,B cell,erythroid progenitor cell,gamma-delta T cell,immature B cell,alpha-beta T cell,immature T cell,natural helper lymphocyte,lymphoblast,B-lymphoblast,blood lymphocyte,monocyte,plasmacytoid dendritic cell,conventional dendritic cell,innate lymphoid cell are subClassOf another slim term!
make: *** [cl.Makefile:288: reports/blood_and_immune_upper_slim.csv] Error 1
```
>* Generate report with:
     - Coverage (% terms in context subsumed by terms in slim, excluding superclasses of slim terms 
        - make this a QC fail if <80%? context ??
     - Number of terms covered by each term in the slim
     - List of terms in context not covered by slim

Example reports can be seen below;
```
96.64%
230
http://purl.obolibrary.org/obo/CL_0002417,primitive erythroid lineage cell
http://purl.obolibrary.org/obo/CL_0000092,osteoclast
http://purl.obolibrary.org/obo/CL_2000074,splenocyte
http://purl.obolibrary.org/obo/FBbt_00005062,prohemocyte
http://purl.obolibrary.org/obo/CL_0000385,prohemocyte (sensu Nematoda and Protostomia)
http://purl.obolibrary.org/obo/CL_0000390,blood cell (sensu Nematoda and Protostomia)
http://purl.obolibrary.org/obo/CL_0000588,odontoclast
http://purl.obolibrary.org/obo/CL_0000779,multinuclear osteoclast
```
```
84.00%
21
http://purl.obolibrary.org/obo/CL_1000432,conjunctival epithelial cell
http://purl.obolibrary.org/obo/CL_1000435,epithelial cell of lacrimal duct
http://purl.obolibrary.org/obo/CL_0000718,compound eye cone cell
http://purl.obolibrary.org/obo/CL_0009001,compound eye retinal cell
```

### CODE REVIEW

**Comment by ubyndr** [comment]
Clearly we need the blood_and_immune_upper_slim and eye_upper_slim first 😄 
image

**Comment by bvarner-ebi** [comment]
THanks, @ubyndr,
@anitacaron, is https://github.com/obophenotype/cell-ontology/pull/1917 OK to be merged?

**Comment by bvarner-ebi** [comment]
> Clearly we need the blood_and_immune_upper_slim and eye_upper_slim first 😄

@ubyndr, #1917 has been merged. Let me know if you need something else to proceed.

File: src/scripts/generic_coverage.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    Args:
+        ontology (str): Ontology name, can be either "uberon" or "all"
+        scope_term (str): Scope of the slim
+
+    Returns:
+        output (str): Multicellular anatomical structure SPARQL query
+    """
+    # ?mas rdfs:isDefinedBy 'obo:uberon.owl'
+    filter_str = "FILTER(strstarts(str(?mas),'http://purl.obolibrary.org/obo/UBERON_'))" if ontology == 'uberon' else ""
```
Comment (anitacaron) [inline src/scripts/generic_coverage.py:97]: You can also use the annotation `rdfs:isDefinedBy` here.

Comment (anitacaron) [reply src/scripts/generic_coverage.py:97]: But `?mas` is already defined by CL in the query.

File: src/scripts/generic_coverage.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        PREFIX UBERON: <http://purl.obolibrary.org/obo/UBERON_>
+        PREFIX CL: <http://purl.obolibrary.org/obo/CL_>
+        SELECT ?mas ?label
+        WHERE
+        {{
+          ?mas <http://www.w3.org/2000/01/rdf-schema#subClassOf>|<http://purl.obolibrary.org/obo/BFO_0000050> {_scope} .
+          ?mas rdfs:isDefinedBy <http://purl.obolibrary.org/obo/cl.owl> .
+          ?mas rdfs:label ?label. 
+          ?s <http://www.w3.org/2000/01/rdf-schema#subClassOf>|<http://purl.obolibrary.org/obo/BFO_0000050> ?mas.
+          FILTER(?s != ?mas)
```
Comment (anitacaron) [inline src/scripts/generic_coverage.py:111]: Why do you need to include this?

**Review by anitacaron — CHANGES_REQUESTED** [review]
The SPARQL queries seem to me all similar. We can work on this together.

File: src/scripts/generic_coverage.py (hunk 1 above)
Comment (ubyndr) [reply src/scripts/generic_coverage.py:97]: We don't use the ontology parameter anymore, I guess I should have removed this line and the parameter...

Comment (anitacaron) [reply src/scripts/generic_coverage.py:97]: Rename the variables in the SPARQL query would be good too.

File: src/scripts/generic_coverage.py (hunk 2 above)
Comment (ubyndr) [reply src/scripts/generic_coverage.py]: I think we needed it back when we were calculating the organ term coverage in CAP, but I cannot justify the reason at the moment. I've just removed it for now

File: src/scripts/generic_coverage.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            {{
+              ?term <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?super. ?super rdfs:label ?label. 
+              ?super <http://www.w3.org/2000/01/rdf-schema#subClassOf>|<http://purl.obolibrary.org/obo/BFO_0000050> {_scope}
+              VALUES ?term {{{' '.join(term_iri_list)}}}
+            FILTER(?term != ?super)
+            }}
+        """
+
+
+def get_term_leaves_list_query(term_iri_list: List[str], scope_term: str) -> str:
```
Comment (anitacaron) [inline src/scripts/generic_coverage.py:136]: Why do you need this query? Isn't this already get in the `get_scope_query`?

**Comment by bvarner-ebi** [comment]
> @bvarner-ebi, Maybe you need to check the slim terms. The requested QC is failing:
>> Fail test if any term in upper_slim is a subclass of another term in the same upper_slim; (enforce naming convention? regex .+_upper_slim$ ???)

Is any overlap forbidden? I understand why it should be limited, but should all overlap be removed?

**Comment by dosumis** [comment]
> Is any overlap forbidden? I understand why it should be limited, but should all overlap be removed?

I'm not totally sure.   Overlap can be a problem for some use cases - e.g. if you wanted to generate a pie chart of cell numbers you'd want a non-overlapping set of terms.  However, it may be hard to balance with getting good coverage.

In this case, I'd be interested to see what happens if you try to get rid of all overlap - would the coverage still be acceptable?

It looks to me like a lot might be fixed by removing  general classes:
'professional antigen presenting cell', 'mononuclear cell'
& then some more specific ones:  immature B-cell, 'B-lymphoblast' 
not sure about this pair 

erythroid progenitor cell | hematopoietic lineage restricted progenitor cell
-- | --

**Comment by anitacaron** [comment]
I removed all overlapping classes, and the coverage is still 97.96%.

**Comment by bvarner-ebi** [comment]
Anything I need to update at this point, @anitacaron?

**Comment by anitacaron** [comment]
@bvarner-ebi, I updated the template. You can check the number of terms in each slim term [before](https://api.triplydb.com/s/xank3692p) and [after](https://api.triplydb.com/s/ED-Xw4ry1).

**Comment by bvarner-ebi** [comment]
> @bvarner-ebi, I updated the template. You can check the number of terms in each slim term [before](https://api.triplydb.com/s/xank3692p) and [after](https://api.triplydb.com/s/ED-Xw4ry1).

Thanks, @anitacaron!
Can you confirm that the deleted terms will still be accessible in the slim? 
For example, macrophage and B cell, two of the terms that have been removed, will still be present in the slim?

**Comment by anitacaron** [comment]
Yes, they'll be. [Here](https://api.triplydb.com/s/F34QgkXtv) is a list of all terms in the slim.

**Review by bvarner-ebi — APPROVED** [review]

**Review by anitacaron — APPROVED** [review]

File: src/scripts/generic_coverage.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    term_leaves_dict = get_term_leaves(list(term_dict.values()), scope)
+    invalid_slim_term_list = get_invalid_subclass_list(list(term_dict.values()))
+    if invalid_slim_term_list:
+        invalid_report_file = output_file.replace("templates/", "reports/overlapping_terms_", 1)
+        with open(invalid_report_file, 'w+', newline='') as invalid_file:
+            write = csv.writer(invalid_file)
+            write.writerows(invalid_slim_term_list)
+        # Disabling the exception for now
+        # raise Exception(f"{file_name} is invalid! {invalid_report_file} report for more details")
```
Comment (ubyndr) [inline src/scripts/generic_coverage.py:207]: @dosumis this part is still generating the overlap report

**Review by anitacaron — APPROVED** [review]

**Comment by ubyndr** [comment]
I don't have authorization to merge this PR. So if everything looks good, can someone with the authorization do the merge?

**Pull request merged**